identity: Add GSSAPI suport for FreeIPA authentication

JAORMX opened the pull request:
##### SUMMARY
This enables the usage of GSSAPI for authentication, instead of having
to pass the username and password as part of the playbook run.

If there is GSSAPI support, this makes the password optional, and will
be able to use the KRB5_CLIENT_KTNAME or the KRB5CCNAME environment
variables; which are standard when using kerberos authentication.

Note that this depends on the urllib_gssapi library, and will only
enable this if that library is available.

This was done since it enables us better security, as the usage of GSSAPI doesn't pass passwords around, and can instead rely on kerberos keytabs or caches.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
identity

JAORMX committed:
identity: Add GSSAPI suport for FreeIPA authentication

This enables the usage of GSSAPI for authentication, instead of having
to pass the username and password as part of the playbook run.

If there is GSSAPI support, this makes the password optional, and will
be able to use the KRB5_CLIENT_KTNAME or the KRB5CCNAME environment
variables; which are standard when using kerberos authentication.

Note that this depends on the urllib_gssapi library, and will only
enable this if that library is available.

abbra commented on lib/ansible/module_utils/ipa.py:
```diff
@@ -68,23 +71,38 @@ def get_json_url(self):
-            self._fail('login', to_native(e))
+        if 'KRB5_CLIENT_KTNAME' in os.environ and HAS_GSSAPI:
+            ccache = "MEMORY:" + str(uuid.uuid4())
+            os.environ['KRB5CCNAME'] = ccache
```
Please do not override existing KRB5CCNAME. You should only use MEMORY: ccaches if no KRB5CCNAME is defined in the environment already.

JAORMX replied:
Good call! Let me update this.

JAORMX replied:
Updated it. Does this make sense?

abbra commented on lib/ansible/module_utils/ipa.py:
```diff
@@ -68,23 +71,38 @@ def get_json_url(self):
+            if not password:
+                self._fail('login', 'Password is required if not using '
+                           'GSSAPI. To use GSSAPI, please set the '
+                           'KRB5_CLIENT_KTNAME or KRB5CCNAME environment '
```
".. or both" -- KRB5_CLIENT_KTNAME only says where to take a keytab from while KRB5CCNAME is used to store currently active tickets. Both can be set by a user because they serve different purpose.

abbra commented:
Looks good to me.

abbra commented:
Tests failed as fedora repositories weren't accessible.

abbra commented:
shipit

JAORMX committed:
identity: Add documentation for GSSAPI authentication for FreeIPA

This documentation describes how to use GSSAPI authentication with the
IPA identity modules.

JAORMX committed:
identity: Add changelog for GSSAPI support for IPA

This adds the changelog entry for the GSSAPI authentication feature for
the IPA identity module.

jamescassell commented:
Awesome!  I'd love to see this in other modules, such as vmware_*, as the underlying pyvmomi supports it.

Akasurde commented:
@jamescassell https://github.com/ansible/ansible/issues/52180
